llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -9119,6 +9120,28 @@ bool llvm::matchSimpleRecurrence(const BinaryOperator *I, PHINode *&P,
   return P && matchSimpleRecurrence(P, BO, Start, Step) && BO == I;
 }
 
+bool llvm::matchSimpleBinaryIntrinsicRecurrence(const IntrinsicInst *I,
+                                                PHINode *&P, Value *&Init,
+                                                Value *&Invariant) {
+  IntrinsicInst *II = nullptr;
+  P = dyn_cast<PHINode>(I->getArgOperand(0));
+  if (!P)
+    P = dyn_cast<PHINode>(I->getArgOperand(1));
+
+  return P &&
+         matchTwoInputRecurrence<IntrinsicInst>(
+             P, II, Init, Invariant,
+             [](const IntrinsicInst *I) {
+               Intrinsic::ID IID = I->getIntrinsicID();
+               return isa<SaturatingInst, MinMaxIntrinsic>(I) ||
+                      IID == Intrinsic::minnum || IID == Intrinsic::maxnum ||
+                      IID == Intrinsic::minimum || IID == Intrinsic::maximum ||
+                      IID == Intrinsic::minimumnum ||
+                      IID == Intrinsic::maximumnum;
+             }) &&
```

Comment:
I would expect that the caller is responsible for checking the intrinsic ID. We should only check that it's a two argument intrinsic (maybe with same arg and ret types) here.